src/core/common/time_ticker.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a time ticker.
+ */
+
+#include "time_ticker.hpp"
+
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "thread/mle_router.hpp"
+
+namespace ot {
+
+TimeTicker::TimeTicker(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mReceivers(0)
+    , mTimer(aInstance, HandleTimer, this)
+{
+}
+
+void TimeTicker::RegisterReceiver(Receiver aReceiver)
+{
+    mReceivers |= Mask(aReceiver);
+
+    if (!mTimer.IsRunning())
+    {
+        mTimer.Start(kTickInterval);
+    }
+}
+
+void TimeTicker::UnregisterReceiver(Receiver aReceiver)
+{
+    mReceivers &= ~Mask(aReceiver);
+
+    if (mReceivers == 0)
+    {
+        mTimer.Stop();
+    }
+}
+
+void TimeTicker::HandleTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<TimeTicker>().HandleTimer();
+}
+
+void TimeTicker::HandleTimer(void)
+{
+    mTimer.FireAt(mTimer.GetFireTime() + kTickInterval);
```

Comment:
This code might not be working well when there is a huge time increase in the system.  In such case, this code would fire the timer continuously without delay until the fire time catches up with  current system time.

Thus I would suggest:
```suggestion
    Time fireTime = mTimer.GetFireTime() + kTickInterval;
    Time now = TimeMilli::GetNow();
    mTimer.FireAt(fireTime > now ? fireTime : now);
```

Or we can change it to simply:
```suggestion
    mTimer.Start(kTickInterval)
```

Reply:
This way of starting the timer is indeed intentional. I think we want the ticks to be emitted at fixed rate vs fixed delay between ticks.

The tick receivers generally use it for tracking durations or  timeout (in seconds). e.g., expiring address resolver cache entries after some time.

For example, if there is a task/operation that holds the execution for a while (e.g., flash erase), and the raw time value does jump ahead more than a second, I think we do want to emit multiple ticks back to back (since in terms of actual time, really multiple seconds have passed).

BTW, `Timer` does work correctly (fires immediately) if we schedule it with a time instance in the past (relative to `now`). Method `Timer::DoesFireBefore()` handles this.

Reply:
Rechecked. Should be fine since posix is using `CLOCK_MONOTONIC_RAW`.